packages/fxa-auth-server/lib/payments/subscription-reminders.ts

```diff
@@ -0,0 +1,216 @@
+  year: (plan, planLength) =>
+    plan.interval_count >= planLength / DAYS_IN_A_YEAR,
+};
+
+export class SubscriptionReminders {
+  private db: any;
+  private mailer: any;
+  private stripeHelper: StripeHelper;
+
+  constructor(
+    private log: Logger,
+    config: ConfigType,
+    private planLength: number = DEFAULT_PLAN_LENGTH,
+    private reminderLength: number = DEFAULT_REMINDER_LENGTH,
+    db: any,
+    mailer: any
+  ) {
+    this.db = db;
+    this.mailer = mailer;
+    this.stripeHelper = Container.get(StripeHelper);
+  }
+
+  /**
+   * For all possible plan.intervals, determine if the plan is sufficiently
+   * long based on planLength.
+   */
+  private isEligiblePlan(plan: Plan): boolean {
+    for (let [interval, checkFn] of Object.entries(eligiblePlansByInterval)) {
+      if (plan.interval === interval && checkFn(plan, this.planLength)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private async getEligiblePlans(): Promise<Plan[]> {
+    const allPlans = await this.stripeHelper.allPlans();
+    return allPlans.filter((plan) => this.isEligiblePlan(plan));
+  }
+
+  /**
+   * Returns a window of time in seconds [startTimeS, endTimeS)
+   * that is exactly one day, reminderLength days from now in UTC.
+   */
+  private getStartAndEndTimes(reminderLengthMs: number): TimeSpanInS {
+    const reminderDay = new Date(Date.now() + reminderLengthMs);
+    // Get hour 0, minute 0, second 0 for today's date
+    const startingTimestamp = new Date(
+      Date.UTC(
+        reminderDay.getUTCFullYear(),
+        reminderDay.getUTCMonth(),
+        reminderDay.getUTCDate(),
+        0,
+        0,
+        0
+      )
+    );
+    // Get hour 0, minute, 0, second 0 for one day from today's date
+    const endingTimestamp = new Date(startingTimestamp.getTime() + MS_IN_A_DAY);
+
+    const startTimeS = Math.floor(startingTimestamp.getTime() / 1000);
+    const endTimeS = Math.floor(endingTimestamp.getTime() / 1000);
+    return {
+      startTimeS,
+      endTimeS,
+    };
+  }
+
+  private async alreadySentEmail(
+    uid: string,
+    currentPeriodStartMs: number,
+    emailParams: SentEmailParams
+  ) {
+    const emailRecord = await SentEmail.findLatestSentEmailByType(
+      uid,
+      EMAIL_TYPE,
+      emailParams
+    );
+    // This could be the first email for a given subscription or a subsequent one.
+    return !!emailRecord && emailRecord.sentAt > currentPeriodStartMs;
+  }
+
+  private async updateSentEmail(uid: string, emailParams: SentEmailParams) {
+    await SentEmail.createSentEmail(uid, EMAIL_TYPE, emailParams);
+  }
+
+  /**
+   * Send out a renewal reminder email if we haven't already sent one.
+   */
+  async sendSubscriptionRenewalReminderEmail(
+    subscription: Stripe.Subscription
+  ): Promise<Boolean> {
+    const { customer } = subscription;
+    // @ts-ignore
+    const { email } = customer;
+    // @ts-ignore
+    const uid = customer.metadata.userid;
```

Comment:
Nit: assert `customer` to be `Stripe.Customer` instead of using ts-ignore?

Reply:
If I change this to `const customer: Stripe.Customer = subscription.customer;`, I get these TS errors:
```
Type 'string | Customer | DeletedCustomer' is not assignable to type 'Customer'.
Type 'string' is not assignable to type 'Customer'.ts(2322)
```

This worked, so I guess I will go with that to avoid type errors at runtime?:
```ts
const customer: Stripe.Customer | Stripe.DeletedCustomer | string = subscription.customer;
if (typeof customer === "string") {
  return false;
}
if (customer?.deleted) {
  return false;
}
// ...
```

Reply:
Barry and I resolved this sync.